File: Medium/131_palindrome_partitioning/solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s)-1
        while l<=r:
            if s[l] != s[r]:
                return False
            l+=1; r-=1
        return True

    def backtracking(self, s: str, res: List[List[str]], temp: List[str]) -> None:
        if len(s) == 0:
            res.append(temp[:])
            return
        for i in range(len(s)):
            part = s[0:i+1]
            if self.isPalindrome(part):
                temp.append(part)
                self.backtracking(s[i+1:], res, temp)
                temp.pop()

    def partition(self, s: str) -> List[List[str]]:
        res = []
        temp = []

        self.backtracking(s, res, temp)

        return res

File: Medium/131_palindrome_partitioning/solution.cpp
class Solution {
public:
    bool isPalindrome(std::string s) {
        int start = 0;
        int end = s.size() - 1;
        while (start <= end) {
            if (s[start] != s[end]) {
                return false;
            }
            start++; end--;
        }
        return true;
    }

    void backtracking(std::string s, vector<vector<std::string>>& res, vector<std::string>& temp) {
        if (s.size() == 0){
            res.push_back(temp);
            return;
        }
        for (int i=0; i<s.size(); i++) {
            std::string part = s.substr(0, i+1);
            if (isPalindrome(part)) {
                temp.push_back(part);
                backtracking(s.substr(i+1), res, temp);
                temp.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        std::vector<vector<string>> res;
        std::vector<string> temp;
        
        backtracking(s, res, temp);

        return res;
    }
};
